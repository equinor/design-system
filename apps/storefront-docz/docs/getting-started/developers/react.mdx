---
title: Developers
tabs:
  - Introduction
  - CSS framework
  - React
  - Figma
mode: draft
route: /getting-started/developers/react/
---

:warning: Information coming soon
