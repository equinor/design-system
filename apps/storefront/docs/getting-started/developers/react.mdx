---
title: Developers
metaTitle: Developers react
tabs:
  - Introduction
  - CSS framework
  - React
  - Figma
mode: draft
route: /getting-started/developers/react/
---

:warning: Information coming soon
