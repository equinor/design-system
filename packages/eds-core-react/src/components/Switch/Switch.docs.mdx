import { Links, PropsTable, Story } from './../../../.storybook/components'
import { Switch } from '.'

# Switch

Selection controls allow users to select options, make decisions and set preferences.

<Links 
    ariaUrl="https://www.w3.org/WAI/ARIA/apg/patterns/switch/"
    figmaUrl="https://www.figma.com/file/0bGXR2sCwMVSDNyyzu5BXrO5/UI%E2%80%94User-Interface?node-id=18442%3A38197"
    documentationUrl="https://eds.equinor.com/0b0c666ab/p/028476-selection-controls/b/86826a"
    sourceUrl="https://github.com/equinor/design-system/blob/develop/packages/eds-core-react/src/components/Switch/Switch.tsx"
    npmUrl="https://www.npmjs.com/package/@equinor/eds-core-react"
/>

<Story id="inputs-selection-controls-switch--introduction" />
<PropsTable />

## Usage

**Selection controls**

<ul>
    <li>Allow users to select options, make decisions and set preferences.</li>
    <li>Are to be visible and understandable at a quick glance.</li>
    <li>Are always accompanied with a clear label.</li>
</ul>

**Switch**

<ul>
    <li>Toggle a single item on or off.</li>
    <li>Immediately activate or deactivate something.</li>
     <li>Switches make it easy to compare available options.</li>
</ul>

```tsx
import { Switch } from '@equinor/eds-core-react'

<Switch label="Label text" />
```

## Examples

### Default states

<Story id="inputs-selection-controls-switch--default-states" />

### Alternative to label

To comply with accessibility, a `label` is always required on inputs. In some cases though, a visual label is not desirable. <br />
In such cases `aria-label` or `aria-labelledby` should be used.

<Story id="inputs-selection-controls-switch--alternative-to-label" />

### Compact

Compact `Switch` using `EdsProvider`.

<Story id="inputs-selection-controls-switch--compact" />

### Table switch

Example of usage with `Switch` and `aria-label` in tables for accessibility.

<Story id="inputs-selection-controls-switch--table-switch" />