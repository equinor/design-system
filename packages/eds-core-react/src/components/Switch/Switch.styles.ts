import styled, { css } from 'styled-components'
import { outlineTemplate } from '@equinor/eds-utils'

export const BaseInputWrapper = styled.span(
  ({ theme: { width, height } }) => css`
    width: ${width};
    height: ${height};
    border-radius: 50%;
    position: relative;
    grid-area: input;
  `,
)

export const BaseInput = styled.input.attrs(({ type = 'checkbox' }) => ({
  type,
}))(
  ({ disabled, theme }) => css`
    appearance: none;
    width: 100%;
    height: 100%;
    grid-area: input;
    margin: 0;
    position: relative;
    z-index: 1;
    cursor: ${disabled ? 'not-allowed' : 'pointer'};
    &:focus {
      outline: none;
    }
    &[data-focus-visible-added]:focus + span {
      ${outlineTemplate(theme.states.focus.outline)}
    }
    &:focus-visible + span {
      ${outlineTemplate(theme.states.focus.outline)}
    }
  `,
)

export const GridWrapper = styled.span`
  display: inline-grid;
  vertical-align: middle;
  grid: [input] 1fr / [input] 1fr;
  place-items: center;
  isolation: isolate;
`
