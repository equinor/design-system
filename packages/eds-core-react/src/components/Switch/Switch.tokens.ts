import { ComponentToken, tokens } from '@equinor/eds-tokens'

const {
  colors: {
    ui: {
      background__medium: { rgba: backgroundColorMedium },
    },
    interactive: {
      primary__hover_alt: { rgba: primaryHoverAlt },
      primary__hover: { rgba: primaryHover },
      primary__selected_highlight: { rgba: activeColor },
      primary__resting: { rgba: primaryResting },
      disabled__fill: { rgba: backgroundDisabled },
      focus: { rgba: focusOutlineColor },
      icon_on_interactive_colors: { rgba: iconInteractive },
    },
    text: {
      static_icons__tertiary: { rgba: iconTertiary },
    },
  },
  typography: {
    navigation: { menu_title: labelTypography },
  },
  clickbounds,
  elevation: { raised: boxShadow },
  spacings: {
    comfortable: { medium_small },
  },
  shape: {
    circle,
    _modes: {
      compact: { icon_button: compactIconButton },
    },
  },
  interactions: {
    focused: { width: focusOutlineWidth },
  },
} = tokens

export type SwitchToken = ComponentToken

export const comfortable: SwitchToken = {
  typography: labelTypography,
  height: circle.minHeight,
  width: circle.minWidth,
  clickbound: {
    height: clickbounds.default__base,
    width: clickbounds.default__base,
    offset: {
      top: `${
        (parseInt(clickbounds.default__base) - parseInt(circle.minHeight)) / 2
      }px`,
      left: `${
        (parseInt(clickbounds.default__base) - parseInt(circle.minWidth)) / 2
      }px`,
    },
  },
  entities: {
    label: {
      spacings: {
        left: medium_small,
      },
    },
    track: {
      width: '32px',
      height: '8px',
      background: backgroundColorMedium,
      border: {
        type: 'border',
        radius: '4px',
      },
      states: {
        active: {
          background: activeColor,
        },
        hover: {
          background: primaryHover,
        },
      },
    },
    handle: {
      height: '16px',
      width: '16px',
      boxShadow,
      background: iconTertiary,
      states: {
        active: {
          background: primaryResting,
        },
        hover: {
          background: primaryHoverAlt,
        },
      },
    },
  },
  modes: {
    compact: {
      height: compactIconButton.minHeight,
      width: compactIconButton.minWidth,
      clickbound: {
        height: clickbounds.compact__standard,
        width: clickbounds.compact__standard,
        offset: {
          top: `${
            (parseInt(clickbounds.compact__standard) -
              parseInt(compactIconButton.minHeight)) /
            2
          }px`,
          left: `${
            (parseInt(clickbounds.compact__standard) -
              parseInt(compactIconButton.minWidth)) /
            2
          }px`,
        },
      },
      entities: {
        track: {
          width: '20px',
          height: '10px',
          states: {
            disabled: {
              background: iconTertiary,
            },
            active: {
              background: primaryResting,
            },
          },
        },
        handle: {
          height: '6px',
          width: '6px',
          background: iconInteractive,
        },
      },
    },
  },
  states: {
    disabled: {
      background: backgroundDisabled,
    },
    focus: {
      outline: {
        width: focusOutlineWidth,
        type: 'outline',
        style: 'dashed',
        color: focusOutlineColor,
      },
    },
    hover: {
      background: primaryHoverAlt,
      entities: {
        handle: {
          background: primaryHover,
        },
      },
    },
  },
}
