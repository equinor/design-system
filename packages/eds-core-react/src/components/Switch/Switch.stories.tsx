import { useState, useEffect, ChangeEvent } from 'react'
import { Switch, SwitchProps, EdsProvider, Table, Density } from '../..'
import styled from 'styled-components'
import { ComponentMeta, Story } from '@storybook/react'
import { data } from '../../stories/data'
import page from './Switch.docs.mdx'

const UnstyledList = styled.ul`
  margin: 0;
  padding: 0;
  list-style-type: none;
`

export default {
  title: 'Inputs/Selection Controls/Switch',
  component: Switch,
  parameters: {
    docs: {
      page,
    },
  },
} as ComponentMeta<typeof Switch>

export const Introduction: Story<SwitchProps> = (args) => (
  <Switch label="Play with me" {...args} />
)

export const DefaultStates: Story<SwitchProps> = () => {
  const [check, setCheck] = useState(false)

  return (
    <UnstyledList>
      <li>
        <Switch label="I'm default off" />
      </li>
      <li>
        <Switch label="I'm default on" defaultChecked />
      </li>
      <li>
        <Switch disabled label="You can't turn me on!" />
      </li>
      <li>
        <Switch disabled defaultChecked label="You can't turn me off!" />
      </li>
      <li>
        <Switch
          onChange={(e: ChangeEvent<HTMLInputElement>) => {
            setCheck(e.target.checked)
          }}
          checked={check}
          label={`Slider is ${check ? 'checked' : 'unchecked'}`}
        />
      </li>
    </UnstyledList>
  )
}
DefaultStates.storyName = 'Default states'

export const AlternativeToLabel: Story<SwitchProps> = () => (
  <Switch aria-label="This label is invisible, but read by screen-readers" />
)
AlternativeToLabel.storyName = 'Alternative to label'

export const Compact: Story<SwitchProps> = () => {
  const [density, setDensity] = useState<Density>('comfortable')

  useEffect(() => {
    // Simulate user change
    setDensity('compact')
  }, [density])

  return (
    <EdsProvider density={density}>
      <UnstyledList>
        <li>
          <Switch label="I'm default off" size="small" />
        </li>
        <li>
          <Switch label="I'm default on" defaultChecked size="small" />
        </li>
        <li>
          <Switch disabled label="You can't turn me on!" size="small" />
        </li>
        <li>
          <Switch
            disabled
            defaultChecked
            label="You can't turn me off!"
            size="small"
          />
        </li>
      </UnstyledList>
    </EdsProvider>
  )
}

export const TableSwitch: Story<SwitchProps> = () => {
  return (
    <Table>
      <Table.Head>
        <Table.Row>
          <Table.Cell>Selected</Table.Cell>
        </Table.Row>
      </Table.Head>
      <Table.Body>
        {data.map((data) => (
          <Table.Row key={data.number}>
            <Table.Cell>
              <Switch aria-label={`Select ${data.description}`} />
            </Table.Cell>
          </Table.Row>
        ))}
      </Table.Body>
    </Table>
  )
}
TableSwitch.storyName = 'Table switch'
