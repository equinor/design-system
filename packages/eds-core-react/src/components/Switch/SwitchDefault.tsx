import { forwardRef, InputHTMLAttributes } from 'react'
import styled, { css } from 'styled-components'
import { bordersTemplate } from '@equinor/eds-utils'
import { BaseInput, BaseInputWrapper, GridWrapper } from './Switch.styles'

type StyledProps = { $isDisabled: boolean }

const Input = styled(BaseInput)(
  ({ disabled, theme }) => css`
    width: ${theme.clickbound.height};
    aspect-ratio: 1/1;
    /*  Track */
    &:checked + span > span {
      background-color: ${disabled
        ? theme.states.disabled.background
        : theme.entities.track.states.active.background};
    }
    /* Handle */
    &:checked + span > span:last-child {
      transform: translate(105%, -50%);
      background-color: ${disabled
        ? theme.states.disabled.background
        : theme.entities.handle.states.active.background};
    }
    &:hover + span {
      background-color: ${disabled
        ? 'transparent'
        : theme.states.hover.background};
    }
    &:hover + span > span:last-child {
      background-color: ${disabled
        ? theme.states.disabled.background
        : theme.states.hover.entities.handle.background};
    }
  `,
)

const Track = styled.span<StyledProps>(
  ({
    $isDisabled,
    theme: {
      states,
      entities: { track },
    },
  }) => css`
    ${bordersTemplate(track.border)}
    border: none;

    width: ${track.width};
    height: ${track.height};
    background-color: ${track.background};
    position: absolute;
    left: 50%;
    top: 50%;
    transform: translate(-50%, -50%);
    transition: background 0.36s;
    ${$isDisabled && {
      backgroundColor: states.disabled.background,
    }}
  `,
)
const Handle = styled.span<StyledProps>(
  ({
    $isDisabled,
    theme: {
      states,
      entities: { handle },
    },
  }) => css`
    background-color: ${handle.background};
    ${$isDisabled && {
      backgroundColor: states.disabled.background,
    }}
    box-shadow: ${handle.boxShadow};
    width: ${handle.width};
    height: ${handle.height};
    border-radius: 50%;
    display: inline-block;
    position: absolute;
    top: 50%;
    transform: translate(0, -50%);
    left: 4px;
    transition: transform 0.36s cubic-bezier(0.78, 0.14, 0.15, 0.86);
  `,
)

type SwitchDefaultProps = {
  disabled?: boolean
} & InputHTMLAttributes<HTMLInputElement>

export const SwitchDefault = forwardRef<HTMLInputElement, SwitchDefaultProps>(
  function SwitchDefault({ disabled, className, style, ...rest }, ref) {
    return (
      <GridWrapper className={className} style={style}>
        <Input {...rest} ref={ref} disabled={disabled} />
        <BaseInputWrapper>
          <Track $isDisabled={disabled} />
          <Handle $isDisabled={disabled} />
        </BaseInputWrapper>
      </GridWrapper>
    )
  },
)

SwitchDefault.displayName = 'SwitchDefault'
