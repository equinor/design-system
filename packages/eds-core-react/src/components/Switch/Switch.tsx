import { forwardRef, InputHTMLAttributes } from 'react'
import styled, { ThemeProvider, css } from 'styled-components'
import { SwitchSmall } from './SwitchSmall'
import { SwitchDefault } from './SwitchDefault'
import { comfortable as tokens } from './Switch.tokens'
import { typographyTemplate, useToken } from '@equinor/eds-utils'
import { useEds } from '../EdsProvider'

type StyledProps = { isDisabled: boolean; size: 'default' | 'small' }

const StyledLabel = styled.label<StyledProps>`
  cursor: ${({ isDisabled }) => (isDisabled ? 'not-allowed' : 'pointer')};
  border: none;
  background-color: transparent;
  vertical-align: middle;
  display: inline-flex;
  align-items: center;
  position: relative;
  gap: ${({ size }) => (size === 'small' ? '12px' : '8px')};
`

const Label = styled.span(
  ({ theme }) => css`
    ${typographyTemplate(theme.typography)};
  `,
)

export type SwitchProps = {
  /** Label for the switch. Required to make it a11y compliant */
  label?: string
  /** Switch size, use the small version with caution */
  size?: 'default' | 'small'
  /** If true, the switch will be disabled */
  disabled?: boolean
} & Omit<InputHTMLAttributes<HTMLInputElement>, 'size'>

export const Switch = forwardRef<HTMLInputElement, SwitchProps>(function Switch(
  { size = 'default', disabled, label, className, style, ...rest },
  ref,
) {
  const { density } = useEds()
  // TODO Temporary workaround untill we can deprecate "size" property (controlled by EdsProvider in the future)
  const overrideDensity = size === 'small' ? 'compact' : density
  const token = useToken({ density: overrideDensity }, tokens)

  return (
    <ThemeProvider theme={token}>
      {label ? (
        <StyledLabel
          size={size}
          isDisabled={disabled}
          className={className}
          style={style}
        >
          {size === 'small' ? (
            <SwitchSmall disabled={disabled} {...rest} ref={ref} />
          ) : (
            <SwitchDefault disabled={disabled} {...rest} ref={ref} />
          )}
          {label && <Label>{label}</Label>}
        </StyledLabel>
      ) : size === 'small' ? (
        <SwitchSmall
          disabled={disabled}
          className={className}
          style={style}
          {...rest}
          ref={ref}
        />
      ) : (
        <SwitchDefault
          disabled={disabled}
          className={className}
          style={style}
          {...rest}
          ref={ref}
        />
      )}
    </ThemeProvider>
  )
})

Switch.displayName = 'Switch'
