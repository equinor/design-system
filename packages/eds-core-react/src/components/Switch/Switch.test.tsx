/* eslint-disable no-undef */

import { render, fireEvent, screen } from '@testing-library/react'
import { axe } from 'jest-axe'
import userEvent from '@testing-library/user-event'
import styled from 'styled-components'

import { Switch } from './Switch'

const StyledSwitch = styled(Switch)`
  clip-path: unset;
`

describe('Switch', () => {
  it('Matches snapshot', () => {
    const { asFragment } = render(<Switch label="switch" />)
    expect(asFragment()).toMatchSnapshot()
  })
  it('Should pass a11y test', async () => {
    const { container } = render(<Switch label="switch" />)
    expect(await axe(container)).toHaveNoViolations()
  })
  it('Should pass a11y test with aria-label', async () => {
    const { container } = render(<Switch aria-label="switch" />)
    expect(await axe(container)).toHaveNoViolations()
  })
  it('Can extend the css for the component', () => {
    const { container } = render(<StyledSwitch label="switch-test" />)
    // eslint-disable-next-line testing-library/no-node-access
    expect(container.firstChild).toHaveStyleRule('clip-path', 'unset')
  })
  it('Has provided label', () => {
    const label = 'Switch label'
    render(<Switch label={label} />)
    const inputNode = screen.getByLabelText(label)
    expect(inputNode).toBeDefined()
  })
  it('Can be turned on and off', () => {
    const labelText = 'Switch label'
    render(<Switch label={labelText} />)
    const switchElement = screen.getByLabelText(labelText)
    expect(switchElement).not.toBeChecked()
    fireEvent.click(switchElement)
    expect(switchElement).toBeChecked()
    fireEvent.click(switchElement)
    expect(switchElement).not.toBeChecked()
  })
  it('Has a small version', () => {
    const labelText = 'Small switch'
    render(<Switch label={labelText} size="small" />)
    const smallSwitch = screen.getByLabelText(labelText)
    expect(smallSwitch).toBeDefined()
  })
  it('Can be disabled', async () => {
    render(
      <div>
        <Switch label="Checkbox one" disabled />
      </div>,
    )
    const one = screen.getByLabelText('Checkbox one')
    expect(one).not.toBeChecked()
    await userEvent.click(one)
    expect(one).not.toBeChecked()
  })
  it('Can be set as default on without being a controlled component', () => {
    const labelText = 'Default checked switch'
    render(<Switch label={labelText} defaultChecked />)
    const switchElement = screen.getByLabelText(labelText)
    expect(switchElement).toBeChecked()
  })
})
