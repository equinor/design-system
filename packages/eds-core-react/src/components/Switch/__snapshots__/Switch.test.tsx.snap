// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Switch Matches snapshot 1`] = `
<DocumentFragment>
  .c4 {
  width: 40px;
  height: 40px;
  border-radius: 50%;
  position: relative;
  grid-area: input;
}

.c2 {
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  width: 100%;
  height: 100%;
  grid-area: input;
  margin: 0;
  position: relative;
  z-index: 1;
  cursor: pointer;
}

.c2:focus {
  outline: none;
}

.c2[data-focus-visible-added]:focus + span {
  outline: 2px dashed var(--eds_interactive__focus,rgba(0,112,121,1));
  outline-offset: 0px;
}

.c2:focus-visible + span {
  outline: 2px dashed var(--eds_interactive__focus,rgba(0,112,121,1));
  outline-offset: 0px;
}

.c1 {
  display: inline-grid;
  vertical-align: middle;
  grid: [input] 1fr / [input] 1fr;
  place-items: center;
  isolation: isolate;
}

.c3 {
  width: 48px;
  aspect-ratio: 1/1;
}

.c3:checked + span > span {
  background-color: var(--eds_interactive_primary__selected_highlight,rgba(230,250,236,1));
}

.c3:checked + span > span:last-child {
  -webkit-transform: translate(105%,-50%);
  -ms-transform: translate(105%,-50%);
  transform: translate(105%,-50%);
  background-color: var(--eds_interactive_primary__resting,rgba(0,112,121,1));
}

.c3:hover + span {
  background-color: var(--eds_interactive_primary__hover_alt,rgba(222,237,238,1));
}

.c3:hover + span > span:last-child {
  background-color: var(--eds_interactive_primary__hover,rgba(0,79,85,1));
}

.c5 {
  border-radius: 4px;
  border: none;
  width: 32px;
  height: 8px;
  background-color: var(--eds_ui_background__medium,rgba(220,220,220,1));
  position: absolute;
  left: 50%;
  top: 50%;
  -webkit-transform: translate(-50%,-50%);
  -ms-transform: translate(-50%,-50%);
  transform: translate(-50%,-50%);
  -webkit-transition: background 0.36s;
  transition: background 0.36s;
}

.c6 {
  background-color: var(--eds_text__static_icons__tertiary,rgba(111,111,111,1));
  box-shadow: 0 1px 5px rgba(0,0,0,0.2),0 3px 4px rgba(0,0,0,0.12),0 2px 4px rgba(0,0,0,0.14);
  width: 16px;
  height: 16px;
  border-radius: 50%;
  display: inline-block;
  position: absolute;
  top: 50%;
  -webkit-transform: translate(0,-50%);
  -ms-transform: translate(0,-50%);
  transform: translate(0,-50%);
  left: 4px;
  -webkit-transition: -webkit-transform 0.36s cubic-bezier(0.78,0.14,0.15,0.86);
  -webkit-transition: transform 0.36s cubic-bezier(0.78,0.14,0.15,0.86);
  transition: transform 0.36s cubic-bezier(0.78,0.14,0.15,0.86);
}

.c0 {
  cursor: pointer;
  border: none;
  background-color: transparent;
  vertical-align: middle;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  position: relative;
  gap: 8px;
}

.c7 {
  margin: 0;
  color: var(--eds_navigation__menu_title_color,rgba(61,61,61,1));
  font-family: Equinor;
  font-size: 1.000rem;
  font-weight: 400;
  line-height: 1.000em;
  -webkit-letter-spacing: 0.013em;
  -moz-letter-spacing: 0.013em;
  -ms-letter-spacing: 0.013em;
  letter-spacing: 0.013em;
  text-align: left;
}

<label
    class="c0"
  >
    <span
      class="c1"
    >
      <input
        class="c2 c3"
        type="checkbox"
      />
      <span
        class="c4"
      >
        <span
          class="c5"
        />
        <span
          class="c6"
        />
      </span>
    </span>
    <span
      class="c7"
    >
      switch
    </span>
  </label>
</DocumentFragment>
`;
